File: 4.semester/IJA/src/application/Connection.java
/* 
 * Project:     Implementing Block Editor in JavaFX
 * Course:      IJA 2018
 * Team name:   Tym xzubri00
 * Members: 
 *      Liscinsky Matus xlisci02
 *      Zubrik Tomas    xzubri00
 */
package application;
import java.util.Vector;

public class Connection {
    Port output;
    Port input;
    Link link;

    /**
     * Constructor of Connection
     * @param  output Output port
     * @param  input  Input port
     */
    public Connection(Port output, Port input) {
        this.output=output;
        this.input=input;

        if(this.input == null || this.output == null){
            System.exit(1);
        }
        
        this.input.connected_to=this.output;
    }

    /**
     * Function check compatibility of connected ports
     * @return 0 if ports are compatible, -1 otherwise
     */
    public int checkCompatibility(){

        Vector<String> keys1 = this.output.keys;
        Vector<String> keys2 = this.input.keys;

        if(keys1.size() == keys2.size())
        	if(keys1.equals(keys2))
                return 0;
        return -1;
    }

    
}

File: 4.semester/ICP/src/connection.cpp
/**
* @file connection.cpp
* @brief Connection implementation
* @author Matúš Liščinský xlisci02, Marcin Vladimír xmarci10
*/
#include "connection.h"

/**
 * @brief Connection constructor
 * @param output source port of connection
 * @param input destination port of connection
 */
Connection::Connection(Port * output, Port * input)
{
    this->output=output;
    this->input=input;

    if(this->input == NULL || this->output == NULL){
        fprintf(stderr,"NULL dereference.");
        exit(1);
    }

    this->input->connected_to=this->output;
}

/**
 * @brief Check compatiblity of connection (compare types of ports in connection)
 * @return if types are identical return 0 else return -1
 */
int Connection::checkCompatibility()
{
    std::vector<std::string> keys1 = this->output->keys;
    std::vector<std::string> keys2 = this->input->keys;

    if(keys1.size() == keys2.size())
        if(std::equal(keys1.begin(), keys1.end(), keys2.begin()))
            return 0;
    return -1;
}

